internal/configv2/configv2.go

```diff
@@ -0,0 +1,511 @@
+	errDuplicateProcessorName
+	errDuplicatePipelineName
+	errMissingPipelines
+	errPipelineMustHaveReceiver
+	errPipelineMustHaveExporter
+	errPipelineMustHaveProcessors
+	errPipelineReceiverNotExists
+	errPipelineProcessorNotExists
+	errPipelineExporterNotExists
+	errMetricPipelineCannotHaveProcessors
+	errUnmarshalError
+)
+
+type configError struct {
+	msg  string          // human readable error message.
+	code configErrorCode // internal error code.
+}
+
+func (e *configError) Error() string {
+	return e.msg
+}
+
+// receiversKeyName is the configuration key name for receivers section.
+const receiversKeyName = "receivers"
+
+// exportersKeyName is the configuration key name for exporters section.
+const exportersKeyName = "exporters"
+
+// processorsKeyName is the configuration key name for processors section.
+const processorsKeyName = "processors"
+
+// pipelinesKeyName is the configuration key name for pipelines section.
+const pipelinesKeyName = "pipelines"
+
+// typeAndNameSeparator is the separator that is used between type and name in type/name composite keys.
+const typeAndNameSeparator = "/"
+
+const (
+	tracesDataTypeStr  = "traces"
+	metricsDataTypeStr = "metrics"
+)
+
+// Load loads a ConfigV2 from Viper.
+func Load(v *viper.Viper) (*configmodels.ConfigV2, error) {
+
+	var config configmodels.ConfigV2
+
+	// Load the config.
+
+	receivers, err := loadReceivers(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Receivers = receivers
+
+	exporters, err := loadExporters(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Exporters = exporters
+
+	processors, err := loadProcessors(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Processors = processors
+
+	pipelines, err := loadPipelines(v)
+	if err != nil {
+		return nil, err
+	}
+	config.Pipelines = pipelines
+
+	// Config is loaded. Now validate it.
+
+	if err := validateConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+// decodeTypeAndName decodes a key in type[/name] format into type and fullName.
+// fullName is the key normalized such that type and name components have spaces trimmed.
+// The "type" part must be present, the forward slash and "name" are optional.
+func decodeTypeAndName(key string) (typeStr, fullName string, err error) {
+	items := strings.SplitN(key, typeAndNameSeparator, 2)
+	if len(items) < 1 {
+		// We need at least one part (the type).
+		err = errors.New("Invalid key")
+		return
+	}
+
+	typeStr = strings.TrimSpace(items[0])
+	if typeStr == "" {
+		// Type must be present.
+		err = errors.New("Invalid key")
+		return
+	}
+
+	var nameSuffix string
+	if len(items) > 1 {
+		// "name" part is present.
+		nameSuffix = strings.TrimSpace(items[1])
+		if nameSuffix == "" {
+			err = errors.New("Invalid key")
+			return
+		}
+	} else {
+		nameSuffix = ""
+	}
+
+	// Create normalized fullName.
+	if nameSuffix == "" {
+		fullName = typeStr
+	} else {
+		fullName = typeStr + typeAndNameSeparator + nameSuffix
+	}
+
+	err = nil
```

Comment:
Hmm, seems pretty idiomatic to me, e.g.: https://tour.golang.org/basics/7

Reply:
naked return yes, use of named return values too, the call `err = nil` that is kind of strange to me since it wasn't set to any other value above if it reaches this point of the code

Reply:
The alternate is to use unnamed return `return fullName, typeStr, err` however I find this error-prone because I can equally easy make a mistake and write `return typeStr, fullName, err`. The compiler will not complain but it is completely wrong. For this reason I prefer to use named returns in scenarios when I have more than one return parameter with the same type (like here). 

This may look slightly unconventional but it is safer.